python/lsst/daf/butler/tests/hybrid_butler.py

```diff
@@ -0,0 +1,431 @@
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+from collections.abc import Collection, Iterable, Mapping, Sequence
+from contextlib import AbstractContextManager
+from typing import Any, TextIO, cast
+
+from lsst.resources import ResourcePath, ResourcePathExpression
+
+from .._butler import Butler
+from .._dataset_existence import DatasetExistence
+from .._dataset_ref import DatasetId, DatasetRef
+from .._dataset_type import DatasetType
+from .._deferredDatasetHandle import DeferredDatasetHandle
+from .._file_dataset import FileDataset
+from .._limited_butler import LimitedButler
+from .._query import Query
+from .._storage_class import StorageClass
+from .._timespan import Timespan
+from ..datastore import DatasetRefURIs
+from ..dimensions import DataCoordinate, DataId, DimensionGroup, DimensionRecord, DimensionUniverse
+from ..registry import CollectionArgType, Registry
+from ..remote_butler import RemoteButler
+from ..transfers import RepoExportContext
+
+
+class HybridButler(Butler):
+    """A `Butler` that delegates methods to internal RemoteButler and
+    DirectButler instances.  Intended to allow testing of RemoteButler before
+    its implementation is complete, by delegating unsupported methods to
+    DirectButler.
+    """
+
+    _remote_butler: RemoteButler
+    _direct_butler: Butler
+
+    def __new__(cls, remote_butler: RemoteButler, direct_butler: Butler) -> "HybridButler":
+        self = cast(HybridButler, super().__new__(cls))
+        self._remote_butler = remote_butler
+        self._direct_butler = direct_butler
+        self._datastore = direct_butler._datastore
+        return self
+
+    def isWriteable(self) -> bool:
+        return self._remote_butler.isWriteable()
+
+    def _caching_context(self) -> AbstractContextManager[None]:
+        return self._direct_butler._caching_context()
+
+    def transaction(self) -> AbstractContextManager[None]:
+        return self._direct_butler.transaction()
+
+    def put(
+        self,
+        obj: Any,
+        datasetRefOrType: DatasetRef | DatasetType | str,
+        /,
+        dataId: DataId | None = None,
+        *,
+        run: str | None = None,
+        **kwargs: Any,
+    ) -> DatasetRef:
+        return self._direct_butler.put(obj, datasetRefOrType, dataId, run=run, **kwargs)
+
+    def getDeferred(
+        self,
+        datasetRefOrType: DatasetRef | DatasetType | str,
+        /,
+        dataId: DataId | None = None,
+        *,
+        parameters: dict | None = None,
+        collections: Any = None,
+        storageClass: str | StorageClass | None = None,
+        **kwargs: Any,
+    ) -> DeferredDatasetHandle:
+        return self._direct_butler.getDeferred(
```

Comment:
`getDeferred` in theory is simple to implement in remote butler since it is a deferred get and get works already. The minor complication is that the direct butler implementation does a check to make sure the get is likely to work (calling datastore at the moment but it should really call `self.exists()` and `self.stored()`) before it returns the deferred handle (this is so that you don't spend hours doing some processing only to find that the get fails). This might be something to add fairly soon (not on this ticket) along with `exists` and `stored`.